From abe098748f5d6d5ecddfea5de4e1018bffb66f1c Mon Sep 17 00:00:00 2001
From: Dong Aisheng <aisheng.dong@nxp.com>
Date: Thu, 7 Sep 2017 14:47:08 +0800
Subject: [PATCH 3331/5242] MLK-17491-37 clk: imx: clk-composite: using kernel
 fractional divider instead of our own

commit  e926ed86ac46283e2100d7ce21a127f67ec7565f from
https://source.codeaurora.org/external/imx/linux-imx.git

Kernel already supports fractional divider.
See: drivers/clk/clk-fractional-divider.c

After patch: ("clk: fractional-divider: add CLK_FRAC_DIVIDER_ZERO_BASED
flag support"), it can supports ZERO based dividers now which be used by
IMX ULP.

Cc: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Dong Aisheng <aisheng.dong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/imx/clk-composite.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/imx/clk-composite.c b/drivers/clk/imx/clk-composite.c
index 7482bb9..c0fc16e 100644
--- a/drivers/clk/imx/clk-composite.c
+++ b/drivers/clk/imx/clk-composite.c
@@ -26,7 +26,7 @@ struct clk *imx_clk_composite(const char *name, const char **parent_name,
 {
 	struct clk_gate *gate = NULL;
 	struct clk_mux *mux = NULL;
-	struct clk_frac_divider *div = NULL;
+	struct clk_fractional_divider *div = NULL;
 	struct clk_hw *mux_hw = NULL, *div_hw = NULL, *gate_hw = NULL;
 	struct clk *clk;
 
@@ -61,6 +61,7 @@ struct clk *imx_clk_composite(const char *name, const char **parent_name,
 		div->nshift = 0;
 		div->nwidth = 3;
 		div->nmask  = 0x7;
+		div->flags = CLK_FRAC_DIVIDER_ZERO_BASED;
 	}
 
 	if (gate_present) {
@@ -80,7 +81,7 @@ struct clk *imx_clk_composite(const char *name, const char **parent_name,
 
 	/* register the composite clk itself */
 	clk = clk_register_composite(NULL, name, parent_name, num_parents,
-				mux_hw, &clk_mux_ops, div_hw, &clk_frac_divider_ops,
+				mux_hw, &clk_mux_ops, div_hw, &clk_fractional_divider_ops,
 				gate_hw, &clk_gate_ops, CLK_SET_RATE_GATE | CLK_SET_PARENT_GATE);
 
 	if (IS_ERR(clk)) {
-- 
1.7.9.5

